### ISSUE TITLE
Navigator User Agent Leak buildID, OS-Cpu, platform, and Iframe

### ISSUE THREAD

**Issue by panderlad** [kkapsner/CanvasBlocker#363]
The useragent faking doesn't work as expected. It leaks the real BuildID of the browser, and all the real browser information on Iframe.

## Description

When tested on https://canvasblocker.kkapsner.de/test/navigatorTest.php this is the results:

Navigator test
Tests the navigator properties.
server site user agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4; rv:53.0) Chrome/55.0.2924.129 Safari/608.18
user agent equal between server and client: true
appCodeName: Mozilla
appName: Netscape
appVersion: 5.0 (Windows)
buildID: 20190521153532
doNotTrack: unspecified
language: en-US
oscpu: Windows NT 6.1; Win64; x64
platform: Win64
product: Gecko
productSub: 20100101
userAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4; rv:53.0) Chrome/55.0.2924.129 Safari/608.18
vendor:
vendorSub:

Values in iFrame
appCodeName@iframe: Mozilla
appName@iframe: Netscape
appVersion@iframe: 5.0 (Windows)
buildID@iframe: 20190521153532
doNotTrack@iframe: unspecified
language@iframe: en-US
oscpu@iframe: Windows NT 6.1; Win64; x64
platform@iframe: Win64
product@iframe: Gecko
productSub@iframe: 20100101
userAgent@iframe: Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:56.0) Gecko/20100101 Firefox/56.0 Waterfox/56.2.10
vendor@iframe:
vendorSub@iframe: 

## Expected Behavior

Should fake all the information and not leak real information.

/
Please fix it. Thanks.

**Comment by kkapsner** [comment]
This problem was solved with version 0.5.9 of CB. Please update it. I cannot reproduce the problem with Waterfox.

If you already have the latest version (0.5.10) and the issue is still present I can reopen it.

**Comment by panderlad** [comment]
I  have the latest version (0.5.10).

Also the extention doesn't work on Comodo Icedrago browser. Can install it but doesn't fake anything.

**Comment by kkapsner** [comment]
That's weird. Also on Comodoe icedragon browser (actually never heard from it before) it works fine for me.

What are your actual CB settings?

Are there any CB related errors in the console (Ctrl + Shift + K or Ctrl + Shift + J)?

**Comment by rlesch** [comment]
Maybe unrelated maybe not: also using Waterfox 56.2.10, and the only way to change the Useragent seems to change/set a 'general.useragent.override' key in about:config and assign an useragent to it. They key gets deleted on every restart, at least i havent found a way to make that change permanent.

Any recent addon that changes useragent fails to do so in Waterfox (and Canvas BLocker is no exeptiopn here. :/ )  Setting eg predefinded Chrome in the navigator options doesnt change anything. UA still the same waterfox one regardless what i set there.
 
The only one works is the outdated  "user agent switcher" that uses aboves method and has teh same probem to keep the change permanent. For me it seems that the Waterfoc useragent is kinda hardcoded.

**Comment by kkapsner** [comment]
The actual server sees the modified userAgent. I have to look into the proxy issue.

**Comment by kkapsner** [comment]
@panderlad Please answer my questions above. Otherwise I'm not able to help you and solve this issue. On my end everything works as designed.

**Comment by kkapsner** [comment]
@rlesch what type of proxy are you using?

**Comment by rlesch** [comment]
local / lan proxy usually for domain filtering or debugging.
eg Privoxy, Proxomitron or Fiddler.

header scan always shows the true useragent.

**Comment by kkapsner** [comment]
I see the problem for the https tunnels but not if direct http is intercepted. So it's a problem with the proxy protocol which I cannot alter. I don't think it's worth opening a feature request for that at Mozilla as you should only use a proxy you trust.
So you the user agent is exposed to the proxy (which you trust and which might need to have that information to work properly... not that deep into it) but not the the real server (you can see it under ```server site user agent:``` in my test page) which we do not trust. So we do not have a problem here.

@panderlad If the problem still exists and you can provide the required information I can reopen this issue.

**Comment by ilikenwf** [comment]
I ran into this with the current version of waterfox-current (68 ESR, essentially).

After playing with the settings buttons, I found that I had to get all the fields needed enabled to make things work. Seems that you have to have certain fields in place and set or it doesn't spoof at all.

These settings did not spoof properly/at all:
```	"protectNavigator": true,
	"navigatorDetails": {
		"userAgent": "Mozilla/5.0 (Windows NT 10.0; rv:{firefoxVersion}) Gecko/20100101 Firefox/{firefoxVersion}",
		"platform": "Windows NT 6.1",
		"oscpu": "Win32",
		"appVersion": "5.0 (Windows)",
		"buildID": "20100101",
		"productSub": "20100101"
	},```

These settings did:
```	"protectNavigator": true,
	"navigatorDetails": {
		"userAgent": "Mozilla/5.0 ({platformDetails}; rv:{firefoxVersion}) Gecko/20100101 Firefox/{firefoxVersion}",
		"platform": "Win32",
		"oscpu": "{platformDetails}",
		"appVersion": "5.0 ({windowManager})",
		"buildID": "20100101",
		"productSub": "20100101",
		"product": "Gecko",
		"osPreset": "Windows",
		"windowManager": "Windows",
		"platformDetails": "Windows NT 10.0",
		"browserPreset": "Firefox",
		"firefoxVersion": "68.0"
	},```

**Comment by kkapsner** [comment]
> These settings did not spoof properly/at all:

Not sure what you exactly mean by that. Only the properties that you specify are spoofed.
You also have to specify the variable `firefoxVersion` in your upper example as you use it in `userAgent` - otherwise there will be an `undefined`.

Can you please provide a screenshot of what you think is not correct?

**Comment by ilikenwf** [comment]
On the test page, oscpu etc etc were not being changed, only the useragent sent to the server...so values I did have set were not changing.

After clicking the "windows" button and the "firefox" button and then filling in my desired values, the navigator values were all finally spoofed properly. The first example, even the navigator values I had set were not being spoofed on your test page: https://canvasblocker.kkapsner.de/test/navigatorTest.php

I'll try to remember to back up my current config, and test the "broken" config again...I'm not necessarily blaming the plugin, but something finally happened after my messing with it that it started working after adding the windows and firefox specific fields.
